docs/architecture/adr-007-address-generation-of-chain-link-proof.md

```diff
@@ -0,0 +1,79 @@
+# ADR 006: Address generation of chain link proof
+
+## Changelog
+
+- January 20th, 2022: Initial draft;
+
+## Status
+
+PROPOSED
+
+## Abstract
+
+Currently, Desmos chain link supports `Bech32`, `Base58` and `Hex` address formats. Besides, each of them only couples one address generation algorithms. 
+Since Desmos will support more addresses from other chains, which might apply the supported address format but with the unsupported algorithm, we SHOULD split the address algorithm from `AddressData` instances.
+
+## Context
+
+The `x/profiles` module gives users the possibility to link their profile to different external accounts.
+However, each of supported address format are sticked on only one generation algorithm now, says:
+1. `Bech32Address` utilizes basic cosmos generation functions which are `ripemd160(sha256(32 bytes public key))[:20]` for simple account and `sha256(aminoCdc.Marshal(multisig public key))[:20]` for multisig account
+2. `Base58Address` generates address in the solana way which is from 32 bytes public key directly
+3. `HexAddress` uses the ethereum generation function keccak64(64 bytes public key)[12:]
+
+Due to the fact all the `AddressData` instances highly coupled to the address generation function, it is impossible to link the supported address format with the different address generation algorithms.
+
+## Decision
+
+We propose to apply `Strategy` design pattern by a new interface to handle address generation from public key, which includes `GenerateAddress` function.
+```go
+type AddressGenerationStrategy interface{
+    GenerateAddress(pk PubKey) []byte
+}
+```
+
+Subsequently, the `AddressData` instance like `Bech32Address` can include the any value of the strategy instance which implements `AddressGenerationStrategy` as a property.
+```proto
+message Bech32Address{
+    ...
+    google.protobuf.Any strategy = 3 [(cosmos_proto.accepts_interface) = "AddressGenerationStrategy"];
+}
+```
```

Comment:
Honestly I'm not a big fan if this approach. We should use an `interface` (and the related serialization strategy - `Any`) only when there are different implementations that need different data to be created. This was true for `AddressData`: 
- `Bech32AddressData` requires a prefix and a value 
- `HexAddressData` requires only a value
- `Base58AddressData` requires only a value

In this case, each address generation strategy does not require anything else aside from the public key, which is passed to the `GenerateAddress` method itself. For this reason, I think the best alternative is to simply add a new method to the `AddressData` interface instead: 
```go
// AddressData is an interface representing a generic external chain address
type AddressData interface {
    proto.Message

    // Validate checks the validity of the AddressData
    Validate() error

    // GetValue returns the address value
    GetValue() string

    // GetAddressGenerationStrategy returns the address generation strategy to be used
    GetAddressGenerationStrategy() AddressGenerationStrategy

    // VerifyAddress verifies that the given address is associated with this address data
    VerifyAddress(address []byte) (bool, error)
}
```

As you can see, I've also changed `VerifyPubKey` to `VerifyAddress`. Now the flow should become: 
1. generate the proper address using the associated strategy;
2. verify that the address matches the address data value.

```go
address, err := GenerateAddress(b.GetAddressGenerationStrategy(), key)
if err != nil {
    return err
}

valid, err := addressData.VerifyAddress(bz)
if err != nil {
    return err
}
```

The `GenerateAddress` is going to be a public method that simply takes an `AddressGenerationStrategy`, a public key and generates the address properly: 
```go
func GenerateAddress(strategy AddressGenerationStrategy, pubKey cryptotypes.PubKey) (address []byte, err error) {
    switch strategy { 
        case StrategyRipemd160:
            return generateRipemd160Address(pubKey)
        // ...
    }
```

This implementation will also require to change all the `AddressData` instances constructors to allow passing an (optional) address generation strategy

Reply:
I'll wait for this to be included in the ADR and then review it. Notify me when ready @dadamu.